res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1378 @@
+            field.mapped_callback(value, co_group_name, field.name);
+        }
+    // No callback, no light!
+    }
+};
+
+TraktorS2MK2.lightDeck = function(group) {
+    // Freeze the lights while we do this update so we don't spam HID.
+    this.batching_LED_update = true;
+    for (var packet_name in this.controller.OutputPackets) {
+        packet = this.controller.OutputPackets[packet_name];
+        TraktorS2MK2.lightGroup(packet, group, group);
+        // These outputs show state managed by this script and do not react to ControlObject changes,
+        // so manually set them here.
+        TraktorS2MK2.outputCallback(0, group, "!shift");
+        TraktorS2MK2.outputCallback(0, group, "!flux_button");
+        TraktorS2MK2.outputCallback(0, group, "!remix_button");
+    }
+
+    this.batching_LED_update = false;
+    // And now send them all.
+    for (packet_name in this.controller.OutputPackets) {
+        var packet_ob = this.controller.OutputPackets[packet_name];
+        packet_ob.send();
+    }
+};
+
+TraktorS2MK2.init = function(id) {
+    TraktorS2MK2.registerInputPackets();
+    TraktorS2MK2.registerOutputPackets();
+    //   var data = [0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x00, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f];
+    //   controller.send(data, data.length, 0x80);
+
+    TraktorS2MK2.controller.setOutput("[Master]", "!usblight", 0x7F, true);
+    TraktorS2MK2.lightDeck("[Channel1]");
+    TraktorS2MK2.lightDeck("[Channel2]");
+    TraktorS2MK2.lightDeck("[EffectRack1_EffectUnit1]");
+    TraktorS2MK2.lightDeck("[EffectRack1_EffectUnit2]");
+};
+
+TraktorS2MK2.shutdown = function() {
+    var data = [];
+    for (var i = 0; i < 38; i++) {
+        data[i] = 0;
+    }
+    // Leave USB plug indicator light on.
+    data[0x1C] = ButtonBrightnessOn;
+    controller.send(data, data.length, 0x80);
+
+    for (i = 0; i < 33; i++) {
+        data[i] = 0;
+    }
+    controller.send(data, data.length, 0x81);
+};
+
+TraktorS2MK2.incomingData = function(data, length) {
+    TraktorS2MK2.controller.parsePacket(data, length);
+};
+
+// The short message handles buttons and jog wheels.
+TraktorS2MK2.shortMessageCallback = function(packet, data) {
+    for (name in data) {
+        field = data[name];
+        if (field.name === "!jog_wheel") {
+            TraktorS2MK2.controller.processControl(field);
+            continue;
+        }
+
+        TraktorS2MK2.controller.processButton(field);
+    }
+};
+
+// There are no buttons handled by the long message, so this is a little simpler.
+TraktorS2MK2.longMessageCallback = function(packet, data) {
+    for (name in data) {
+        field = data[name];
+        TraktorS2MK2.controller.processControl(field);
+    }
+};
+
+TraktorS2MK2.samplerGainKnob = function(field) {
+    for (var i = 1; i <= 8; i++) {
```

Comment:
please choose either "<" or "<=" for i=1 loops so it's consistent

Reply:
done 4ab004c